examples/ctc.dx

```diff
@@ -30,115 +33,88 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
-        True  -> logprob_of_first_label
-        False -> log 0.000001
-
-  same_as_last = \ilabels s.
-    o = n_to_i $ ordinal s
-    select (o >= 2) (ilabels.s == ilabels.(clipidx _ (o - 2))) False
-
-  safe_idx = \prev s.
-    select (s >= 0) prev.(clipidx _ s) (log 0.0)
-
-  -- Todo: As suggested by Adam Paske, we could get rid of these
-  -- logaddexp calls with a newtype that overloads + and *
-  update = \t prev.
-    case ordinal t == 0 of
+def ctc {vocab time position} [NonEmpty vocab, NonEmpty position, NonEmpty time, Eq time, Eq vocab]
+    (blank: vocab)
+    (logits: time=>vocab=>(LogSpace Float))
+    (labels: position=>vocab)
+      : LogSpace Float =
+  ilabels = prepend_and_interleave blank labels
+  normalized_logits = for t. normalize logits.t
+
+  -- Initialize dynamic programming table.
+  logprob_start_with_blank = normalized_logits.first_ix.blank
+  logprob_of_first_label   = normalized_logits.first_ix.(labels.first_ix)
+  log_prob_seq_t0 = yield_state (for pos. Exp (-10000.0)) \lpRef.
+    lpRef!(Left  first_ix) := logprob_start_with_blank
+    lpRef!(Right first_ix) := logprob_of_first_label  -- 2nd index
+
+  safe_idx_sub = \prev s y.
+    if (ordinal s) < y
+      then zero
+      else prev.(unsafe_from_ordinal _ (unsafe_nat_diff (ordinal s) y))
+
+  log_prob_seq_final = fold zero \t prev.
+    case t == first_ix of
       True -> log_prob_seq_t0
       False -> for s.
-        cond = ilabels.s == blank || same_as_last ilabels s
-        labar = logaddexp prev.s (safe_idx prev (n_to_i (ordinal s) - 1))
-        other = logaddexp labar  (safe_idx prev (n_to_i (ordinal s) - 2))
-        ans = select cond labar other
-        ans + normalized_logits.t.(ilabels.s)
+        same_as_last = if (ordinal s) >= 2
+          then ilabels.s == ilabels.(unsafe_from_ordinal _ ((ordinal s) -| 2))
+          else False
 
-  log_prob_seq_final = fold log_prob_seq_t0 update
+        -- Equation 6 from CTC paper.
+        ans = if ilabels.s == blank || same_as_last
+          then prev.s + safe_idx_sub prev s 1
+          else prev.s + safe_idx_sub prev s 1 + safe_idx_sub prev s 2
+        ans * normalized_logits.t.(ilabels.s)
 
-  -- Todo: nicer way to get last two elements of log_prob_seq_final.
-  seq_length = 1 + size (position & (Fin 2))
-  endlabel = log_prob_seq_final.((unsafe_nat_diff seq_length 2)@_)
-  endspace = log_prob_seq_final.((unsafe_nat_diff seq_length 1)@_)
-  logaddexp endlabel endspace
+  -- Sum over last two entries in the dynamic programming table.
+  last_label = log_prob_seq_final.(Right (last_ix, False))
+  last_space = log_prob_seq_final.(Right (last_ix, True ))
+  last_label + last_space
 
 
 '## Demo
 
-def randIdxNoZero (n:Type) [Ix n] (k:Key) : n =
-  unif = rand k
-  from_ordinal n $ unsafe_i_to_n $
-    (1 + (f_to_i (floor ( unif * i_to_f ((n_to_i $ size n) - 1)))))
-
-Vocab = Fin 6
-position = Fin 3
-blank = 0@Vocab
+-- Todo: Produce proofs that Fin N is NonEmpty for all N > 0
```

Comment:
I honestly don't see how to do this, so it might be better to leave out the TODO 😕

Reply:
Yeah, the only way I can think of doing this is either in the compiler.  I also didn't end up using the `Subset` constraint because then I'd need instances like `Subset (Fin 10) (Fin 20)`.  Will do.